tests/crypto/src/common_test.c

```diff
@@ -0,0 +1,143 @@
+/*$$$LICENCE_NORDIC_STANDARD<2019>$$$*/
+
+#include <stdio.h>
+#include <stdarg.h>
+#include <string.h>
+#include <stdint.h>
+#include <stddef.h>
+#include <misc/printk.h>
+#include <dk_buttons_and_leds.h>
+#include <drivers/entropy.h>
+#include <mbedtls/ctr_drbg.h>
+
+#include "common_test.h"
+
+/* Entries must correspond to order in enum test_vector_t found in common_test.h */
+const size_t test_vector_sizes[] = {
+	sizeof(test_vector_aes_t),
+	sizeof(test_vector_aead_t),
+	sizeof(test_vector_ecdsa_verify_t),
+	sizeof(test_vector_ecdsa_sign_t),
+	sizeof(test_vector_ecdsa_random_t),
+	sizeof(test_vector_ecdh_t),
+	sizeof(test_vector_hash_t),
+	sizeof(test_vector_hmac_t),
+	sizeof(test_vector_hkdf_t),
+	sizeof(test_vector_ecjpake_t),
+};
+
+/* Entries must correspond to order in enum test_vector_t found in common_test.h */
+const size_t test_vector_name_offset[] = {
+	offsetof(test_vector_aes_t, p_test_vector_name),
+	offsetof(test_vector_aead_t, p_test_vector_name),
+	offsetof(test_vector_ecdsa_verify_t, p_test_vector_name),
+	offsetof(test_vector_ecdsa_sign_t, p_test_vector_name),
+	offsetof(test_vector_ecdsa_random_t, p_test_vector_name),
+	offsetof(test_vector_ecdh_t, p_test_vector_name),
+	offsetof(test_vector_hash_t, p_test_vector_name),
+	offsetof(test_vector_hmac_t, p_test_vector_name),
+	offsetof(test_vector_hkdf_t, p_test_vector_name),
+	offsetof(test_vector_ecjpake_t, p_test_vector_name),
+};
+
+mbedtls_ctr_drbg_context ctr_drbg_ctx;
+
+static int entropy_func(void *ctx, unsigned char *buf, size_t len)
+{
+	return entropy_get_entropy(ctx, buf, len);
+}
+
+int init_ctr_drbg(const unsigned char *p_optional_seed, size_t len)
+{
+	static const unsigned char ncs_seed[] = "ncs_drbg_seed";
+
+	const unsigned char *p_seed;
+
+	if (p_optional_seed == NULL) {
+		p_seed = ncs_seed;
+		len = sizeof(ncs_seed);
+	} else {
+		p_seed = p_optional_seed;
+	}
+
+	struct device *p_device = device_get_binding(CONFIG_ENTROPY_NAME);
+	if (!p_device) {
+		return -ENODEV;
+	}
+
+	mbedtls_ctr_drbg_init(&ctr_drbg_ctx);
+	return mbedtls_ctr_drbg_seed(&ctr_drbg_ctx, entropy_func, p_device, p_seed, len);
+}
+
+uint32_t unhexify(uint8_t *p_output, char const *p_input)
+{
+	unsigned char c, c2;
+
+	if (p_input == NULL) {
+		return 0;
+	}
+
+	if (p_output == NULL) {
+		return 0;
+	}
+
+	int len = strlen(p_input) / 2;
+
+	while (*p_input != 0) {
+		c = *p_input++;
+		if (c >= '0' && c <= '9') {
+			c -= '0';
+		} else if (c >= 'a' && c <= 'f') {
+			c -= 'a' - 10;
+		} else if (c >= 'A' && c <= 'F') {
+			c -= 'A' - 10;
+		} else   {
+			printk("unhexify ERROR");
+		}
+
+		c2 = *p_input++;
+		if (c2 >= '0' && c2 <= '9') {
+			c2 -= '0';
+		} else if (c2 >= 'a' && c2 <= 'f') {
+			c2 -= 'a' - 10;
+		} else if (c2 >= 'A' && c2 <= 'F') {
+			c2 -= 'A' - 10;
+		} else   {
+			printk("unhexify ERROR");
+		}
+
+		*p_output++ = (c << 4) | c2;
+	}
+
+	return len;
+}
```

Comment:
Is it possible to use strtol? I see it's in heavily used in NCS already.

Reply:
The strings are given as e.g.: `"a544218dadd3c10583db49cf39"`. If there was whitespace between each hex byte I could iterate over it using strtol nicely.

An alternative is to move each hex byte into an intermediate buffer and use strtol on that.
I can make this change if desired

Reply:
Ah, I see, I guess it is OK as-is then.

Reply:
But please at least put the hexchar_to_num procedure in a function.

Reply:
Changed in 553abae